lib/puma/cluster.rb

```diff
@@ -254,15 +279,45 @@ def worker(index, master)
       # things in shape before booting the app.
       @launcher.config.run_hooks :before_worker_boot, index, @launcher.events
 
-      server = start_server
+      server = @server ||= start_server
+      restart_server = Queue.new << true << false
+
+      if fork_worker
+        restart_server.clear
+        Signal.trap "SIGCHLD" do
+          Process.wait(-1, Process::WNOHANG) rescue nil
+          wakeup!
+        end
```

Comment:
I found a bug in this code while testing in a production app- if the application spawns a subprocess and waits on its exit status (e.g. via `Process.detach`), it races against the `wait` in this signal handler, so the expected `Process::Status` containing the exit status is intermittently `nil`. I'll make a followup PR with a fix.